services/diagnostic/handlers.go

```diff
@@ -0,0 +1,1222 @@
+	"github.com/influxdata/kapacitor/services/hipchat"
+	"github.com/influxdata/kapacitor/services/httppost"
+	"github.com/influxdata/kapacitor/services/influxdb"
+	"github.com/influxdata/kapacitor/services/k8s"
+	"github.com/influxdata/kapacitor/services/mqtt"
+	"github.com/influxdata/kapacitor/services/opsgenie"
+	"github.com/influxdata/kapacitor/services/pagerduty"
+	"github.com/influxdata/kapacitor/services/pushover"
+	"github.com/influxdata/kapacitor/services/sensu"
+	"github.com/influxdata/kapacitor/services/slack"
+	"github.com/influxdata/kapacitor/services/smtp"
+	"github.com/influxdata/kapacitor/services/snmptrap"
+	"github.com/influxdata/kapacitor/services/swarm"
+	"github.com/influxdata/kapacitor/services/talk"
+	"github.com/influxdata/kapacitor/services/telegram"
+	"github.com/influxdata/kapacitor/services/udp"
+	"github.com/influxdata/kapacitor/services/victorops"
+	plog "github.com/prometheus/common/log"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func Error(l *zap.Logger, msg string, err error, ctx []keyvalue.T) {
+	if len(ctx) == 0 {
+		l.Error(msg, zap.Error(err))
+		return
+	}
+
+	if len(ctx) == 1 {
+		el := ctx[0]
+		l.Error(msg, zap.Error(err), zap.String(el.Key, el.Value))
+		return
+	}
+
+	if len(ctx) == 2 {
+		x := ctx[0]
+		y := ctx[1]
+		l.Error(msg, zap.Error(err), zap.String(x.Key, x.Value), zap.String(y.Key, y.Value))
+		return
+	}
+
+	// This isn't great wrt to allocation, but should not ever actually occur
+	fields := make([]zapcore.Field, len(ctx)+1) // +1 for error
+	fields[0] = zap.Error(err)
+	for i := 1; i < len(fields); i++ {
+		kv := ctx[i-1]
+		fields[i] = zap.String(kv.Key, kv.Value)
+	}
+
+	l.Error(msg, fields...)
+}
+
+// Alert Service Handler
+
+type AlertServiceHandler struct {
+	l *zap.Logger
+}
+
+func zapFieldsFromContext(ctx []keyvalue.T) []zapcore.Field {
+	fields := []zapcore.Field{}
+	for _, kv := range ctx {
+		fields = append(fields, zap.String(kv.Key, kv.Value))
+	}
+
+	return fields
+}
+
+func (h *AlertServiceHandler) WithHandlerContext(ctx ...keyvalue.T) alertservice.HandlerDiagnostic {
+	fields := zapFieldsFromContext(ctx)
+
+	return &AlertServiceHandler{
+		l: h.l.With(fields...),
+	}
+}
+
+func (h *AlertServiceHandler) MigratingHandlerSpecs() {
+	h.l.Debug("migrating old v1.2 handler specs")
+}
+
+func (h *AlertServiceHandler) MigratingOldHandlerSpec(spec string) {
+	h.l.Debug("migrating old handler spec", zap.String("handler", spec))
+}
+
+func (h *AlertServiceHandler) FoundHandlerRows(length int) {
+	h.l.Debug("found handler rows", zap.Int("handler_row_count", length))
+}
+
+func (h *AlertServiceHandler) CreatingNewHandlers(length int) {
+	h.l.Debug("creating new handlers in place of old handlers", zap.Int("handler_row_count", length))
+}
+
+func (h *AlertServiceHandler) FoundNewHandler(key string) {
+	h.l.Debug("found new handler skipping", zap.String("handler", key))
+}
+
+func (h *AlertServiceHandler) Error(msg string, err error, ctx ...keyvalue.T) {
+	if len(ctx) == 0 {
```

Comment:
Is this duplicate from the above Error function?

Reply:
yup. don't know how I missed this.